nav2_dwb_controller/dwb_critics/src/rotate_to_goal.cpp

```diff
@@ -76,16 +85,21 @@ double RotateToGoalCritic::scoreTrajectory(const dwb_msgs::msg::Trajectory2D & t
   }
 
   // If we're sufficiently close to the goal, any transforming velocity is invalid
-  if (fabs(traj.velocity.x) > EPSILON || fabs(traj.velocity.y) > EPSILON) {
+  if (fabs(traj.velocity.x) > EPSILON) {
     throw nav_core2::IllegalTrajectoryException(name_, "Nonrotation command near goal.");
   }
+  if (!ignore_y_velocity_) {
```

Comment:
Why is the `ignore` line necessary? If this is checking if its an inplace rotation or not, there shouldn't be Y velocity to ignore, neither X nor Y should be anything substantive.

Reply:
I guess that depends on the quality of your IMU. Right now, the Turtlebot 3 virtual IMU in Gazebo is really bad. It tells me that we are slipping sideways at up to 0.5 m/s at times. see ROBOTIS-GIT/turtlebot3_simulations#75

I can see that the turtlebot just passes through the values from the Gazebo physics model, and that's as far as I managed to debug. I'm not sure why we get such strange results from the physics model.

Anyhow, it seemed legit to ignore Y velocities on a diff drive robot, so I included this here.

In ROS 1 move_base it avoids this problem by not passing through Y velocities on a diff drive robot. I didn't take that approach because I figured the critics should know there is a real Y velocity

Reply:
Usually when ever I've seen Y velocity in gazebo its been a result of incorrect setups, are you certain its just the IMU plugin? 

We shouldn't be adding things that just satisfy 1 robot's particular simulation parameters and maintain that. In philosophy I think this is trying to mask a problem with the simulation environment with projection controller code

Reply:
Assuming there isn't a bug in the Gazebo physics engine, I suspect the problem has to do with the masses, moments of inertia, and collision boxes in the turtlebot 3 model. The odom data with the noisy velocities comes from the gazebo diff drive plugin, which merely passes through data from the physics model.

How do you deal with noisy IMU data on your robots? Is there a ROS package for filtering the IMU data that we should consider adding to the test configuration, or is the IMU data filtered at a lower level?

Reply:
Assuming this is the issue I've run into before, its not an issue with gazebo, its an issue with the navigation configuration, TF, or model. I'm sorry to be so vague, I probably spent 2 months on a problems like this in college and I just dont remember anymore what it is that I did to resolve it, but there's no IMU on Earth with (reasonable) accuracy that should give you sideways slippage like that amount. That's definitely some frame/model issue unless the variances in the plugin that was set was outrageously large (which again should be fixed there not here). 

If this is through gazebo, you shouldn't need to filter the IMU, you can set the variances to 0 and the data out is perfect. I would recommend trying that experiment actually; setting the IMU and diff drive plugin variances to 0 so its perfect, I bet you'll still see this issue.

Reply:
Second followup: if you're using IMU and wheel encoders in gazebo for state estimation, how are you fusing them?

Reply:
I can reproduce this without any nav2 compononents. All I have to do is load the Turtlebot3 model in gazebo and run teleop. That eliminates the navigation configuration. TF is completely unused in this configuration, so I think we can eliminate that as well.

The problem shows itself in the odom topic which is published by the gazebo_diff_drive plugin. Looking through the code, this doesn't do anything sophisticated like emulate wheel encoders or even add noise to the output. It just gets the current velocity value from the physics engine and sticks it in the odom message. So I think we can eliminate noise models as well.

I'm pretty sure the only things left that can go wrong are the model and gazebo. Like you say, it is probably the model, but I don't see anything obviously wrong. I'll admit I haven't weighed, measured and computed inertia matrices for each of the components that makes up the model. That's the next step.

In the meantime, Turtlebot3 is about the only robot ported to ROS2. Without these workarounds, nobody can really use Nav 2. What if I add another PR to undo these Y velocity workarounds? I'll just leave it sitting in the PR queue until this simulation bug is fixed or there is another robot to work with that doesn't have this problem, but it will be ready to go in at the touch of the "Squash and merge" button.

Reply:
Noise is added in the URDF under the gazebo plugin. https://github.com/ROBOTIS-GIT/turtlebot3/blob/master/turtlebot3_description/urdf/turtlebot3_waffle.gazebo.xacro#L98

If you see that happen without navigation2, then the issue is definitely TF or model (where TF, I mean frame definitions, by model, I mean gazebo plugin configuration). The weight/inertia probably dont have anything to do with it, its most likely the setup of the diff drive plugin, if you're saying you're just looking at the output  from that plugin, the IMU has nothing to do with it. The only 2 places to look at are the URDF frames and the diff drive plugin configuration.

Reply:
Given what I know about the turtlebot3 repos maintained by ROBOTIS, I'd make a _really_ good bet they messed up the plugin, I've found a number of fundamental issues in their packages

Reply:
This is an example of a configuration I know works, based on a robot I use at home
```
    <gazebo>
      <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
        <alwaysOn>true</alwaysOn>
        <legacyMode>false</legacyMode>
        <publishWheelTF>true</publishWheelTF>
        <publishWheelJointState>true</publishWheelJointState>
        <odometrySource>world</odometrySource>
        <publishTf>true</publishTf>
        <wheelAcceleration>0</wheelAcceleration> 
        <wheelTorque>5</wheelTorque>
        <updateRate>60</updateRate>
        <leftJoint>wheel_left_joint</leftJoint>
        <rightJoint>wheel_right_joint</rightJoint>
        <wheelSeparation>.230</wheelSeparation>
        <wheelDiameter>0.070</wheelDiameter>
        <torque>1.</torque>
        <commandTopic>raw_cmd_vel</commandTopic>
        <odometryTopic>odom</odometryTopic>
        <odometryFrame>odom</odometryFrame>
        <robotBaseFrame>base_footprint</robotBaseFrame>
        <rosDebugLevel>na</rosDebugLevel>
      </plugin>
    </gazebo>
```

So if you try that, and it works, then its that, otherwise its frames.

Reply:
I want to say I recall something about transforming frames that happens in the backend of gazebo as a result of poorly defined frames in the simulation case for transforming odometry data from the encoders into the world frame (as both these configurations do) and then the Y has some contribution even though there shouldn't be.

Reply:
I've reverted the Y velocity workaround for now. 

Things have changed in the ros2 version of the plugin such that your example configuration no longer works, as is. I'll spend some time this week trying to get it working.